engineering-plans/0004-subgraph-grafting.md

```diff
@@ -0,0 +1,143 @@
+           probably be faster to copy the data with `create table new.table
+           as table old.table`; that requires that we manually create
+           constraints and indexes after copying the data. As that requires
+           a deeper change in the code that creates the database schema for
+           a subgraph, we'll leave that as a possible improvement
+        3. copy dynamic data sources from the source subgraph. The copying
+           will create new `DynamicEthereumContractDataSource` entities,
+           and all their child entities, by creating a new id for the
+           dynamic data source and adjusting the id's of child entities
+           accordingly
+        4. rewind the new subgraph to the desired block by reverting to
+           that block (both data and metadata)
+        5. set the `latestEthereumBlockHash` and
+           `latestEthereumBlockNumber` of the new subgraph's
+           `SubgraphDeployment` to the desired block
+    - For `reuse`:
+        1. change the `deployment` field of the source subgraph's dynamic
+           data sources to point at the new subgraph
+        2. rename the database schema `sgdNNN` for the source subgraph's
+           data to the target subgraph's name using `alter schema sgdOLD
+           rename to sgdNEW`
+        3. rewind the new subgraph to the desired block
+        4. remove the metadata for the old subgraph
+        5. delete the entry for `sgdOLD` from `deployment_schemas` Strictly
+           speaking, we also need to remove the entry for `sgdOLD` from
+           each `graph-node`'s `Store.subgraph_cache` and
+           `Store.storage_cache` but do not have a mechanism in place for
+           that. We should check, but leaving these entries in the cache
+           should be ok; it might lead to less than helpful error messages
+           when trying to query the old subgraph but should cause no harm
+           beyond that
+
+## Tests
+
+Details TBD, but the tests need to cover at least subgraph manifests
+without a `deployment` and each of the supported `modes`
+
+## Migration
+
+This is purely additive and requires no migration.
+
+## Documentation
+
+The `deployment` field in subgraph manifests will be documented [in the
+public
+docs](https://thegraph.com/docs/define-a-subgraph#the-subgraph-manifest)
+and the [manifest reference](https://github.com/graphprotocol/graph-node/blob/master/docs/subgraph-manifest.md)
+
+## Implementation Plan
+
+The plan is described above under 'Implementation'
+
+## Open Questions
+
+- Should subgraphs deployed with `copy` or `reuse` be marked somehow as it
+  will be very hard to recreate them in other `graph-node` installations?
+  Recreating such a subgraph would require deploying the source subgraph,
+  letting it index past the source block and then deploying the target
+  subgraph. Since the source subgraph could itself be the result of
+  grafting, it might be necessary to perform these operations recursively
```

Comment:
Marked as in "marked in the explorer"? Apart from that, wouldn't the grafted deployments be marked in the database via their deployment mode?

Reply:
I wasn't going to store deployment mode explicitly, just extract it from the manifest on creation and act accordingly. We could store it in the database, either in the `SubgraphDeployment` or the `SubgraphManifest`, though it's not clear to me what we need to store it for.